libc/resolve_path.c

```diff
@@ -0,0 +1,488 @@
+	check_and_free_str("/etc", canonicalize_file_name("/etc/../etc/"));
+	check_and_free_str("/etc", canonicalize_file_name("//etc/..//../etc/"));
+}
+
+
+TEST(resolve_path, canonicalize_pwd_simple)
+{
+	/* don't care about current dir */
+	TEST_ASSERT_EQUAL_INT(0, chdir("/etc"));
+
+	check_and_free_str("/etc", canonicalize_file_name("."));
+	check_and_free_str("/", canonicalize_file_name(".."));
+	check_and_free_str("/", canonicalize_file_name("./.."));
+	check_and_free_str("/", canonicalize_file_name("../."));
+
+	check_and_free_str("/etc/passwd", canonicalize_file_name("passwd"));
+	check_and_free_str("/etc/passwd", canonicalize_file_name("./passwd"));
+
+	check_and_free_str("/etc", canonicalize_file_name("../etc"));
+}
+
+
+/* note canonicalize_path_name(path) == realpath(path, NULL) so no need for separate tests */
+TEST(resolve_path, realpath_abs_noalloc)
+{
+	char result[PATH_MAX];
+
+	TEST_ASSERT_EQUAL_STRING("/", realpath("/", result));
+	TEST_ASSERT_EQUAL_STRING("/", realpath("/.", result));
+	TEST_ASSERT_EQUAL_STRING("/", realpath("//", result));
+	TEST_ASSERT_EQUAL_STRING("/", realpath("///", result));
+	TEST_ASSERT_EQUAL_STRING("/", realpath("/./", result));
+
+	TEST_ASSERT_EQUAL_STRING("/", realpath("/..", result));
+	TEST_ASSERT_EQUAL_STRING("/", realpath("/../../.", result));
+
+	TEST_ASSERT_EQUAL_STRING("/etc", realpath("/etc", result));
+	TEST_ASSERT_EQUAL_STRING("/etc", realpath("/etc//", result));
+	TEST_ASSERT_EQUAL_STRING("/etc", realpath("/etc/.", result));
+
+	TEST_ASSERT_EQUAL_STRING("/", realpath("/etc/..", result));
+	TEST_ASSERT_EQUAL_STRING("/", realpath("/etc/../.", result));
+	TEST_ASSERT_EQUAL_STRING("/etc", realpath("/etc/../etc/", result));
+	TEST_ASSERT_EQUAL_STRING("/etc", realpath("//etc/..//../etc/", result));
+}
+
+
+TEST(resolve_path, realpath_pwd_noalloc)
+{
+	char result[PATH_MAX];
+
+	/* don't care about current dir */
+	TEST_ASSERT_EQUAL_INT(0, chdir("/etc"));
+
+	TEST_ASSERT_EQUAL_STRING("/etc", realpath(".", result));
+	TEST_ASSERT_EQUAL_STRING("/", realpath("..", result));
+	TEST_ASSERT_EQUAL_STRING("/", realpath("./..", result));
+	TEST_ASSERT_EQUAL_STRING("/", realpath("../.", result));
+
+	TEST_ASSERT_EQUAL_STRING("/etc/passwd", realpath("passwd", result));
+	TEST_ASSERT_EQUAL_STRING("/etc/passwd", realpath("./passwd", result));
+
+	TEST_ASSERT_EQUAL_STRING("/etc", realpath("../etc", result));
+}
+
+
+/* failures due to different reasons */
+TEST(resolve_path, realpath_errno)
+{
+	char toolong[PATH_MAX + 10];
+
+	check_null_and_errno(EINVAL, realpath(NULL, NULL));
+	check_null_and_errno(ENOENT, realpath("/asdfasdf", NULL));
+
+	memset(toolong, 'a', sizeof(toolong));
+	toolong[0] = '/';
+	toolong[PATH_MAX + 9] = 0;
+
+	check_null_and_errno(ENAMETOOLONG, realpath(toolong, NULL));
+	check_null_and_errno(ENOTDIR, realpath("/etc/passwd/fake", NULL));
+	/* ELOOP tested by symlink_loop */
+}
+
+
+/* FIXME: this will fail due to not respecting NAME_MAX */
```

Comment:
PATH_MAX?

Reply:
no, `NAME_MAX` is the limit of single path node name. `libphoenix` common implementation doesn't enforce this limit (still might be enforced by different filesystems), so this test could pass on our system.

But it fails against `glibc` implementation and making it work seems like too much work (You need to put `/` every `NAME_MAX`, than actually create the resulting path because we don't want `ENOENT` here).